Ppxlib

tatchi opened the pull request:
This is my attempt at addressing https://github.com/ocaml-ppx/ppxlib/issues/143. It would need careful review as I'm definitely not a `ppx` expert. At least the tests are passing which I guess is a good sign 😁

* First commit https://github.com/ocaml-ppx/ppx_import/commit/d8b346c1f26582d155f44ad4e0eede77560e2800 only formats the code (should make it easier to review the rest)
* Second commit https://github.com/ocaml-ppx/ppx_import/commit/233d9d0afa8af0d30fb27f7c40eb9f902997ac56 actually migrate the code to using `ppxlib`

tatchi commented on src/ppx_import.ml:
```diff
@@ -182,311 +164,414 @@ let get_modtype_decl ~loc sig_items parent_lid elem =
+          (* The equivalent of not specifying the variance explicitly.
+             Since the very purpose of ppx_import is to include the full definition,
+             it should always be sufficient to rely on the inferencer to deduce variance. *)
+          (Ppxlib.Asttypes.NoVariance, Ppxlib.Asttypes.NoInjectivity) ))
```
Definitely not sure about that one. Is `Ppxlib.Asttypes.NoInjectivity` correct ?

ejgallego replied:
This is indeed tricky stuff, @gasche do you know who could help with this?

gasche replied:
I just looked at this (sorry for the delay). It's a bit tricky but I think that the current code in the PR is correct:
- for abstract type, the default is "non injective", so `NoInjectivity` is the conservative choice
- non-abstract types are always injective, so you could use `Injective` here, but it looks like the OCaml parser uses `NoInjective` by default as well and lets the type-checker figure things out afterwards.

tatchi commented on src/ppx_import.ml:
```diff
@@ -182,311 +164,414 @@ let get_modtype_decl ~loc sig_items parent_lid elem =
+    | { pexp_loc; _ } ->
+        Location.raise_errorf ~loc:pexp_loc "Invalid [@with] syntax"
+  in
+  let find_attr s attrs =
+    try
+      Some
+        (List.find (fun { attr_name = x; _ } -> x.txt = s) attrs).attr_payload
+    with Not_found -> None
```
I didn't find the `Ast_convenience.find_attr` equivalent so I just inlined the function

ejgallego commented:
Thanks a lot @tatchi ! I will give it some extended testing this week, I have some development which make extensive use of ppx_import

I'm OK with the code formatting, tho I do believe that it'd be better if you can submit it as a separate PR so we can review this one free of formatting changes.

If we are gonna format ppx_import automatically we should have it happen as part of the CI , add the ocamlformat config, and a makefile target `fmt` that invokes `dune fmt`, we should also check that @gasche and @whitequark are Ok with it.

gasche commented:
Those who do the work decide; @ejgallego if you are doing the main work of integrating this nice-sounding PR, you get to decide on the formatting :-)

ejgallego commented:
> Those who do the work decide; @ejgallego if you are doing the main work of integrating this nice-sounding PR, you get to decide on the formatting :-)

I have a similar view, if @tatchi who is writing this PR would like to have the code formatted, that's fine to me.

ejgallego commented:
Needs fixing for OCaml versions < 4.12:
```
File "src/ppx_import.ml", line 193, characters 16-21:
- 193 |       Typ.arrow label lhs (core_type_of_type_expr ~subst rhs)
-                       ^^^^^
- Error: This expression has type Asttypes.arg_label
-        but an expression was expected of type
-          Migrate_parsetree.Ast_412.Asttypes.arg_label
- ```

tatchi commented:
> > Those who do the work decide; @ejgallego if you are doing the main work of integrating this nice-sounding PR, you get to decide on the formatting :-)
> 
> I have a similar view, if @tatchi who is writing this PR would like to have the code formatted, that's fine to me.

The reason why I formatted the code is because I'm not used to doing that manually. Therefore, I knew I was going to auto-format the files changed by this PR, and figured it would be better to have a first commit that only formats the files. That would make the review easier to avoid extra diff related to formatting changes. I can definitely revert my commit and try to format the changes by hand if you prefer 😄

> Needs fixing for OCaml versions < 4.12:
> 
> ```
> File "src/ppx_import.ml", line 193, characters 16-21:
> - 193 |       Typ.arrow label lhs (core_type_of_type_expr ~subst rhs)
> -                       ^^^^^
> - Error: This expression has type Asttypes.arg_label
> -        but an expression was expected of type
> -          Migrate_parsetree.Ast_412.Asttypes.arg_label
> - ```
> ```

I'm a bit lost between all these `ast` conversions 😅 I tried to re-use the [Convert](https://github.com/ocaml-ppx/ppx_import/blob/master/src/ppx_types_migrate.ml#L10) module but it didn't work. After some research, I stumbled upon [this example](https://github.com/ocaml-ppx/ocamlformat/pull/1482/files#diff-2fc6bbdcde4932e902be2368839cf7e82bf172d31ce0608804fb9ecad4b7ba55R28) and used the same method. Everything seems to compile now except for OCaml 4.04.2 due to [dependencies constraints that can't be met](https://travis-ci.org/github/ocaml-ppx/ppx_import/jobs/771341564#L267)

tatchi commented:
Would that make sense to restrict `ppx_deriving >= 5.0` since this is [the version that switched to ppxlib](https://github.com/ocaml-ppx/ppx_deriving/blob/master/CHANGELOG.md#50-26102020)? If so, I guess we would have to drop support for OCaml < 4.05.0 anyway [since previous versions are not compatible with `ppx_deriving >= 5.0`](https://github.com/ocaml-ppx/ppx_deriving/blob/v5.0/ppx_deriving.opam#L17)

gasche commented:
Personally I think that requiring (OCaml >= 4.05) is now very reasonable.
My compass for "which old version should we still support?" is "Debian stable", which currently uses 4.05.0.

ejgallego commented:
> The reason why I formatted the code is because I'm not used to doing that manually. Therefore, I knew I was going to auto-format the files changed by this PR, and figured it would be better to have a first commit that only formats the files. That would make the review easier to avoid extra diff related to formatting changes. I can definitely revert my commit and try to format the changes by hand if you prefer

On the contrary, I'd say let's try to avoid doing more work than what we need. If you pass me your ocamlformat config we can already merge the formatting changes in a separate PR.

gasche commented:
I think that what @ejgallego has in mind is to have a first PR (or himself committing directly) with your pure-reformat change, and a second PR with the actual changes, in the interest of making the "whole diff" review of the PR easier.

tatchi commented:
> I think that what @ejgallego has in mind is to have a first PR (or himself committing directly) with your pure-reformat change, and a second PR with the actual changes, in the interest of making the "whole diff" review of the PR easier.

Sounds good to me, I'll open a PR with the formatting changes 😊

tatchi commented:
I rebased my changes on top of master. Second, commit bump `ocamlformat` to the latest version because the `0.15.0` version was conflicting with `ppxlib`. Fortunately, it didn't introduce too many formatting changes. Let me know though if you prefer to have that as a separate PR.

ejgallego commented:
> I rebased my changes on top of master. The second commit bumps `ocamlformat` to the latest version because the `0.15.0` version was conflicting with `ppxlib`. Fortunately, it didn't introduce too many formatting changes. Let me know though if you prefer to have that as a separate PR.

Thanks @tatchi , I'd prefer to have as a separate PR if you don't mind, so review is easier.

I assume the PR is ready for review, correct? What about the CI failure?

tatchi commented:
> Thanks @tatchi , I'd prefer to have as a separate PR if you don't mind, so review is easier.

Alright, I removed the commit :)
 
> I assume the PR is ready for review, correct? What about the CI failure?

Yes, it should be ready for review. The CI was failing for OCaml `4.04.2`. I removed this version since it won't be supported anymore (see https://github.com/ocaml-ppx/ppx_import/pull/54#issuecomment-842092027) and now it looks ok :)

ejgallego commented:
Great, thanks @tatchi , will review ASAP (cc: @gasche in case he want to do a pass too)

ejgallego commented on ppx_import.opam:
```diff
@@ -13,9 +13,8 @@ tags: [ "syntax" ]
 
 depends: [
   "ocaml"                   {              >= "4.04.2" }
```
This needs update.

tatchi replied:
Good catch! Thanks 😁

ejgallego reviewed (approved):
Looks good to me, thanks!

I will leave a couple of days to get comments, do some more testing on my own, then merge.

gasche commented:
I skimmed the code and it looks overall fine (the port is rather non-invasive, which is good news). There is one question on variance that is subtle, I will look at it later. (It's about moving from one version of the AST to another, rather than ppxlib per se, in a corner case of the type declaration syntax.)

My experience with these changes is that there are some failures that are not caught by running the code, but by trying them in the real world. Stuff like breakage with older OCaml versions, or breakage when people combine it with other ppxes, or use a non-Dune build system. I think that the real test with this PR will be building the reverse dependencies of `ppx_import` on the opam repository. I would encourage @ejgallego to try to do this as soon as possible (after merging the PR), for example (if you don't have a better workflow) by preparing a new ppx_import release on opam-repository, with the idea that we may iterate a bit more to fix issues as they come up.

ejgallego commented:
Note that the opam file seems to have a couple of issues:
- the ocaml lower bound is incorrect
- I am not sure why dune 1.2 is now the lower bound for ppx_import, I think that version won't work actually, the 1.11 bound is hard AFAICT [and in fact we should raise it to 2.0]

@tatchi , if you feel like addressing these issues it is fine, otherwise I'll tweak them myself before the release.

Regarding my testing, I'm getting this problem:
```ocaml
Error: Type unsupported for ppx [of_sexp] conversion
File "serlib/ser_sorts.ml", line 19, characters 2-25:
19 |   [%import: Sorts.family]
```
where the `serlib/ser_sorts.ml` file is:
```ocaml
type family =
  [%import: Sorts.family]
  [@@deriving sexp,yojson]
```
and the original type is:
```ocaml
type family = InSProp | InProp | InSet | InType
```
the build definition for this library is
```lisp
(library
 (name serlib)
 (public_name coq-serapi.serlib)
 (synopsis "Serialization Library for Coq")
 (preprocess (staged_pps ppx_import ppx_sexp_conv ppx_deriving_yojson))
 (libraries coq.stm sexplib))
```
I tried a few tweaks without success, didn't investigate more yet.

This is a showstopper I'm afraid, as this is a basic use case of ppx_import IMHO.

ejgallego commented:
Actually the test suite should catch this, as it doesn't happen with some derivers, but it does happen with others, I'm gonna strengthen the test-suite adding ppx_deriving_sexp so this PR will properly fail in CI.

ejgallego commented:
@tatchi I took the liberty to rebase this branch and push, now that #59 is in master, so CI properly fails with the testing issue we detected.

pitag-ha commented on src/ppx_import.ml:
```diff
@@ -207,41 +197,49 @@ let rec core_type_of_type_expr ~subst type_expr =
       row_fields
       |> List.map (fun (label, row_field) ->
-             let label = Location.mknoloc label in
+             let label = Ocaml_common.Location.mknoloc label in
```
What do you think about inlining a `mknoloc` function instead of using the compiler-libs one?

In that context: it might even be worth filing an issue on ppxlib to add a `mknoloc` function to `Ppxlib.Location`. I think on one hand not having one is intentional so that people are aware that using a ghost location isn't standard, but on the other hand I've already seen several projects where that's necessary.

pitag-ha reviewed (commented):
Thanks a lot also from the ppxlib side, @tatchi!

As you can see, I only have a few questions, mostly concerning the use of compiler-libs, since that's an unstable API. In that regard also out of curiosity: why do you need to make the use of compiler-libs explicit throuout `ppx_import.ml` (by using `Ocaml_common`) now, whereas before it wasn't necessary?

There are still a lot of use cases of compiler-libs (which were already there before the port), but I think that's in the nature of `ppx_import` and will be unavoidable. Just to get a better understanding: from what I can tell `ppx_import` needs the compiler-libs `Env` module to access the definition details and the compiler-libs `Types` module and everything implied by that as a consequence. Is that more or less what happens?

_Update_: The paragraph below refers to [this comment](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r654434232):
About my suggestion about trying ppxlib's `Instrument` API: as I've mentioned, ppxlib doesn't provide another way to specify that `ppx_import` needs to be applied before other rewriters currently. We need to implement a way to make that possible in ppxlib. What do you think about trying out the "register as instrumentation hack", at least to see if with that the new test passes?

Of course, the two last questions are directed both to you and the `ppx_import` folks (cc @ejgallego?)

And one last thing: this port registers `ppx_import` as a whole file transformation, although `ppx-import` is an extension node rewriter. For efficiency, it would definitely better to define it as an extension node rewriter. But `ppx_import`'s syntax makes that difficult in general and impossible at the moment (I can give more detail on that). In the ppxlib team we have an idea about how to add a "virtual extension node" to ppxlib to make that possible. Then, we could also think about to problem about the order in which PPXs are applied in the context of extension node rewriters. The problem is that I'm not sure how long that will take, so it might be good to leave that all for a separate PR.

pitag-ha commented on src/ppx_import.ml:
```diff
@@ -499,18 +516,28 @@ let module_type ~tool_name mapper modtype_decl =
-      let type_declaration = type_declaration ~tool_name in
-      let module_type = module_type ~tool_name in
-      {default_mapper with type_declaration; module_type})
+  Ppxlib.Driver.V2.register_transformation ~rules:[] ~impl:mapper#structure
+    "ppx_import"
```
Sorry, somehow I didn't sent this in the review. See the [general comment of my review](https://github.com/ocaml-ppx/ppx_import/pull/54#pullrequestreview-686155575) for more information.

Of course, this is only meant as a possibility for a temporary hack to make sure that `ppx_import` is applied before other rewriters (such as `ppx_sexp_conv`) are.

```suggestion
  let open Ppxlib.Driver in
  (* Currently, ppxlib only provides a way to specify when a rewriter should be applied
     (with respect to other rewriters), if the rewriter is an instrumentation. *)
  let temporary_hack =
    Instrument.V2.make mapper#structure ~position:Instrument.Before
  in
  V2.register_transformation ~rules:[] ~instrument:temporary_hack "ppx_import"
```

ejgallego replied:
The hack works, thanks a lot!

pitag-ha replied:
Cool! Do you want to merge this like that? Or do you want to wait until in ppxlib we've implemented both a way to register `ppx_import` as extension point and a solution to this problem?

ejgallego replied:
I don't mind following what you think it is best, @tatchi WDYT?

pitag-ha replied:
Hey, to come back to this and sum up the situation: the other PPXs apart from `ppx_import` pretty much all already rely on modern ppxlib (and therefore, in their newest version, are incompatible with current `ppx_import` due to relying on OMP1). This port seems very nice, functional and non-intrusive. I've suggested reducing the use of `compiler-libs` at a few places which could be done without much effort, but that's not a huge problem and it's up to you two. Apart from that the only problems this port has are due to missing features in `ppxlib` and improving them would take time:
- it's written as whole file transformation and therefore takes up one whole AST pass by its own. But that's also the case with the current OMP1 implementation.
- it has to register `ppx_import` as "instrumentation", which isn't great but practically doesn't do any harm.

So I think merging and releasing this approach would be a good idea. The two points above can still be improved in a different PR when `ppxlib` is ready for that. But of course, that's just my opinion. Let's see what @tatchi says.

ejgallego replied:
Indeed, I'm happy to move ahead as you propose, we can wait a bit to see if @tatchi could address your excellent comments, or if it is busy we can indeed do a release with the current PR and hopefully address the comments later.

pitag-ha commented:
> Regarding my testing, I'm getting this problem:
>```
>Error: Type unsupported for ppx [of_sexp] conversion
>File "serlib/ser_sorts.ml", line 19, characters 2-25:
>19 |   [%import: Sorts.family]
>```

To me it seems like this problem is that ppxlib tries to apply `ppx_sexp_conv` before `ppx_import` and therefore fails. I've proposed trying [this](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r654434232) to make sure that's what happens.

> it doesn't happen with some derivers, but it does happen with others

For what kind of derivers does it not happen? Are the ones where it happens maybe the ones that rely on `ppx_import` being applied them before and the ones where it doesn't happen don't rely on that?

pitag-ha commented:
> After some research, I stumbled upon [this example](https://github.com/ocaml-ppx/ocamlformat/pull/1482/files#diff-2fc6bbdcde4932e902be2368839cf7e82bf172d31ce0608804fb9ecad4b7ba55R28) and used the same method.

I haven't found that anywhere in the PR, but just to make sure: @tatchi, are you using that trick in the end? I'm asking, because using that internal ppxlib API would lead to compatibility problems with future ppxlib versions.

tatchi commented:
> > After some research, I stumbled upon [this example](https://github.com/ocaml-ppx/ocamlformat/pull/1482/files#diff-2fc6bbdcde4932e902be2368839cf7e82bf172d31ce0608804fb9ecad4b7ba55R28) and used the same method.
> 
> I haven't found that anywhere in the PR, but just to make sure: @tatchi, are you using that trick in the end? I'm asking, because using that internal ppxlib API would lead to compatibility problems with future ppxlib versions.

Not strictly the same but something very similar that I've used here: https://github.com/ocaml-ppx/ppx_import/pull/54/files#diff-5f520cdcf9624f27ab1dc90a5a6e92042a28d5b52f5cc9036ae60820ee045ca2R2

Looking at the [current unreleased changes from ppxlib](https://github.com/ocaml-ppx/ppxlib/blob/master/CHANGES.md#unreleased), the `Compiler_version` will soon be removed from the public API. Any idea how we could achieve the same without that?

pitag-ha commented on src/ppx_types_migrate.ml:
```diff
@@ -1,29 +1,23 @@
-module Pb = Ast_411.Parsetree
-module Ob = Ast_411.Outcometree
-module IMigrate = Convert (Versions.OCaml_current) (Versions.OCaml_411)
+module Conv = Ppxlib_ast.Select_ast (Ppxlib_ast.Compiler_version)
```
```suggestion
module Conv = Ppxlib_ast.Selected_ast
```

Thanks for pointing me here, @tatchi! Having both the functor `Select_ast` and the module `Selected_ast` in our API is extremely confusing. `Selected_ast` is staying in the API, whereas we're removing `Select_ast`, and also `Compiler_version`, from it. They were not even meant to be there.

tatchi replied:
Oh nice, that should be straightforward then 😁 Thanks again for the review, I'll try to address your comments in the coming days (except for the ones that are because of missing `ppxlib` features)

pitag-ha replied:
Cool, thanks!

tatchi commented on src/ppx_import.ml:
```diff
@@ -95,20 +78,23 @@ let rec try_open_module_type env module_type =
       (string_of_lid lid)
 
 let locate_sig ~loc env lid =
+  let open Ocaml_common in
```
Looks like it's not needed. Good catch! 🤗

tatchi commented on src/ppx_import.ml:
```diff
@@ -156,14 +143,15 @@ let try_get_tsig_item f ~loc:_ sig_items elem =
   loop sig_items
 
 let get_type_decl ~loc sig_items parent_lid elem =
+  let open Ocaml_common in
```
Not needed 😄

tatchi commented on src/ppx_import.ml:
```diff
@@ -440,13 +455,14 @@ let rec psig_of_tsig ~subst (tsig : Compat.signature_item_407 list) =
             ; oval_attributes = [] }
         in
-        let oval = Primitive.print p oval in
+        let oval = Ocaml_common.Primitive.print p oval in
         oval.Ot.oval_prims
```
I'm definitely not familiar with that either 😅 But looking at the [source code](https://github.com/ocaml/ocaml/blob/005ba108dbe9f14a07175481f7aa8501571f96e3/typing/primitive.ml#L159-L164), it can be either `[p.prim_name; p.prim_native_name]` or `[p.prim_name]` depending on whether `p.prim_native_name` is an empty string.

Does that make sense ?

pitag-ha replied:
It does, thanks! What would you think about inlining those 4 lines you've linked to? That way `ppx_import` wouldn't use `Outcometree` anymore.

pitag-ha replied:
Although I have to admit that that only indirectly has to do with the port to `ppxlib` in the sense that one of the advantages `ppxlib` tries to achieve is that PPXs don't need to use the unstable `compiler-libs` anymore.

tatchi replied:
Thanks, that should be good now 🥳💪

tatchi commented on src/ppx_import.ml:
```diff
@@ -345,12 +357,13 @@ let is_self_reference lid =
     |> uncapitalize
```
Definitely makes sense. Thanks!

tatchi commented on ppx_import.opam:
```diff
@@ -13,9 +13,8 @@ tags: [ "syntax" ]
-  "ppx_tools_versioned"     {              >= "5.4.0"  }
-  "ocaml-migrate-parsetree" {              >= "1.7.0"  }
+  "dune"                    {              >= "1.2.0"  }
+  "ppxlib"                  {              >= "0.22.0"  }
```
I'm sorry but I don't remember why I used `0.22.0` as an upper bound (if there was any reason ^^).

tatchi commented on src/ppx_import.ml:
```diff
@@ -207,41 +197,49 @@ let rec core_type_of_type_expr ~subst type_expr =
       row_fields
       |> List.map (fun (label, row_field) ->
-             let label = Location.mknoloc label in
+             let label = Ocaml_common.Location.mknoloc label in
```
May I ask you why does that matter? We still use the `Ocaml_common` module in other places, so I'm not sure to understand.

tatchi replied:
Done in https://github.com/ocaml-ppx/ppx_import/pull/54/commits/40f96e6f22a231d232c31382f61a095279a3e713

If it looks correct, I'll try to open a PR to the `ppxlib` repo 👍

tatchi commented:
Alright, except for the comment I left [here](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r662440920), I should have addressed all your comments. Thanks again for the review @ejgallego and @pitag-ha 🤗

ejgallego commented:
> Alright, except for the comment I left [here](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r662440920), I should have addressed all your comments. Thanks again for the review @ejgallego and @pitag-ha

Thanks! Let me know folks when you want this merged and I will attempt a release.

bbc2 commented:
I tested this branch and I'm afraid there might be an issue with interface files.  When using `[%import: Bar.t]` in a `.mli` file, I get:

```
> dune build @check
File "foo.mli", line 1, characters 11-17:
1 | type t = [%import: Bar.t]
               ^^^^^^
Error: Uninterpreted extension 'import'.
```

It may be related to the change in how this ppx is registered.

I was able to reproduce this in CI with a new test case (`src_test/ppx_deriving/test_ppx_import.mli`):

- Based on the master branch: https://github.com/bbc2/ppx_import/runs/2992436370 (tests fail but in an expected way)
- Based on this branch: https://github.com/bbc2/ppx_import/runs/2992460858 ("Uninterpreted extension 'import'")

I'm quite excited by this change so I went ahead and tested it on our code base.  This looks unexpected so I thought you might want to know.  Hopefully, I'm not missing something obvious.

ejgallego commented:
Thanks a lot @bbc2 , indeed it is great we are strengthening ppx_import's test suite , would you mind submitting the test case as a separate PR?

bbc2 commented:
Sure, here it is: https://github.com/ocaml-ppx/ppx_import/pull/61

pitag-ha commented:
> I'm afraid there might be an issue with interface files.

Thanks for noticing! What's happening is that currently, both with the "normal" registration and with the `Instrument` registration, only the `structure` of the `Ast_traverse` object is passed to the `register` function. To also rewrite interface files, the `signature` of that object also needs to be passed to it:
```
Ppxlib.Driver.V2.register_transformation ~impl:mapper#structure ~intf:mapper#signature "ppx_import"
```
But now that we're using the `Instrument` trick, adding `signature` isn't possible.

So either we have to implement a way in `ppxlib` to say when a rewriter should be applied that rewrites both implementation and interface files or we directly implement all the features needed for `ppx_import` to fit well with `ppxlib`. I'll come back to you on this.

pitag-ha commented:
> I'll come back to you on this.

I've discussed this with @NathanReb (thanks @NathanReb!) and we've come up with a solution that would be great as a long term solution. For that, the `ppx_import` syntax would have to be made conform with the OCaml AST structure. Since breaking changes are always controversial, I've opened a separate [issue](https://github.com/ocaml-ppx/ppx_import/issues/62) to discuss that. Please, have a look at that issue.

To make the port of `ppx_import` to `ppxlib` work, we've also already considered for a while adding a virtual extension node in `ppxlib` that would capture the current `ppx_import` syntax. That solution has two downsides: one, it would add a`ppxlib` layer that's not conform with the OCaml AST just for `ppx_import`; and two, if someone also uses a standard `core type` extension node along with `ppx_import`, we'd have to raise.

So let's first discuss what you think about the issue I've opened. If you think that's a good idea, we could proceed as follows:
1. We make a `ppxlib` release in which we add a "hidden" function `register_ppx_import` just serving the purpose to register `ppx_import` with its current syntax as whole file transformation and making sure it's applied before `sexp` and `yojson` and all.
3. You use that function here in this port and release the port as `1.x`.
4. You make a major release `2.0.0` of `ppx_import` which fixes the `ppx_import` syntax as explained in the issue.
5. We make a PR to register `ppx_import` as extension node rewriter. 

What do you think?

gasche commented:
As someone who only looks at ppx stuff intermittently (when it stops working), some extra explanations would help.

1. What is the current problem with supporting ppx_import as a ppxlib extension? What I understood from [your previous message](https://github.com/ocaml-ppx/ppx_import/pull/54#issuecomment-876213548) is that the (current) ppxlib implementation rewrites structures / ml files, but not signatures / mli files. Is that the problem that makes a ppxlib-based ppx_import currently unsatisfying?
2. How is your suggestion to change the syntax going to help with this problem?
3. When you say that ppx_import does not conform to the OCaml AST, you mean that the interpretation of the extension point depends on its placement in a wider context, and transforms other parts of the program than just the AST node it is placed at. Is it technically impossible to do this with ppxlib, or rather a question of recommandation/taste, or something in between? I would expect that "generating code outside the place where the extension is placed" would be a not-uncommon need among ppx rewriters (for example: the extension runs on an expression, but it generates helper definitions that should be placed at the toplevel of the currently processed module, to be evaluated only once).

gasche commented:
Looking at the [Ppxlib.Driver documentation](https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib/Driver/index.html), my impression is that the issue is that unlike many other transformation passes (`{lint,preprocess,enclose}_{intf,impl}`, etc.), `instrument` only comes with a structure-processing version, and the signature-processing counterpart (`instrument_intf`) is missing. (And the current way

pitag-ha commented:
> As someone who only looks at ppx stuff intermittently (when it stops working), some extra explanations would help.

Sorry for that!

> What is the current problem with supporting ppx_import as a ppxlib extension?

Due to its syntax, `ppx_import` cannot be registered as a normal extension node rewriter. Unless we implement a virtual extension node in `ppxlib` just for `ppx_import`, `ppx_import` with its current syntax has to be registered as a whole file transformation. If you register it as a whole file transformation in the usual way (with `impl`/`intf`), it will be applied after `sexp` etc (see [here](https://github.com/ocaml-ppx/ppxlib/blob/master/src/driver.mli#L115-L124) for order of application) and therefore applying `sexp` will fail. So as a temporary "hack" around that [I suggested](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r654434232) registering it with `instrument` for now until we come up with a long-term solution: `instrument` has a feature to specify when it should be applied. But `instrument` doesn't allow to rewrite interface files. So that's not an option.

> Would a reasonable fix for this be to implement an `instrument_signature` pass?

I would say that would be a "practical" fix, but I'm not sure about "reasonable". `ppx_import` is not an instrumentation. Making use of the existing `instrument` feature was meant as a temporary trick. If we go down the road of adding something to `ppxlib` to make `ppx_import` work as whole file transformation, I would rather add a "hidden" `register_ppx_import` function to the driver.

> How is your suggestion to change the syntax going to help with this problem?

 If `ppx_import` had the syntax I suggested in [this issue](https://github.com/ocaml-ppx/ppx_import/issues/62), it could be registered as extension node, which would be far more efficient and furthermore we wouldn't have this problem of wrong order of application (I've made sure of that in [this quick ppxlib test](https://github.com/pitag-ha/ppxlib/blob/test-derivers-extension-nodes-order/test/driver/order_derivers_extension_notes/run.t#L31-L60)).

About your question 3. [here](https://github.com/ocaml-ppx/ppx_import/pull/54#issuecomment-879343928):
> When you say that ppx_import does not conform to the OCaml AST, you mean that the interpretation of the extension point depends on its placement in a wider context, and transforms other parts of the program than just the AST node it is placed at.

Yes, the end of the sentence captures exactly what I want to say: the expansion of the node spreads onto the rest of the tree. From what I understand, extension nodes are meant to be expanded (as nodes), not to expand the rest of the tree.

> Is it technically impossible to do this with ppxlib, or rather a question of recommandation/taste, or something in between?

It is possible to do that with `ppxlib`, but only if you register your PPX as a whole file transformation. But that's far less efficient and should only be done if necessary: if you register you PPX as whole file transformation, it takes up one whole AST pass on its own; if you register it as context-free rule (e.g. extension node), it is applied together with all other registered context-free rules in one single AST pass. `ppx_import` does not depend on its context by nature but only due to its syntax. With the [proposed syntax](https://github.com/ocaml-ppx/ppx_import/issues/62) it would be context-free. What do you think about discussing the proposed syntax (probably better on the issue)? I'm curious about the pro's for the current syntax and/or con's for the proposed syntax (apart from the breaking change) in general.

gasche commented:
> I would say that would be a "practical" fix, but I'm not sure about "reasonable". ppx_import is not an instrumentation. Making use of the existing instrument feature was meant as a temporary trick. If we go down the road of adding something to ppxlib to make ppx_import work as whole file transformation, I would rather add a "hidden" register_ppx_import function to the driver.

The way I understand things, ppxlib may actually benefit from giving people the ability to order whole-pass transformations. I'm not sure what "an instrumentation" is, but it seems to provide this feature; yet it is limited in an artificial way (why wouldn't instrumentations want to rewrite signature files as well?). Making `instrument` more coherent with other ppxlib passes seems less ad-hoc, and not harder to do, than adding a "hidden register_ppx_import" function, and it may be of benefit to other users of ppxlib. Is there a reason why that would be a bad idea? Could it not be helpful to help transition other existing extensions?

> What do you think about discussing the proposed syntax (probably better on the issue)?

Yes, we can discuss the proposed syntax. However in principle I'm not so satisfied with the idea that we have to break backward-compatibility of the extension to be able to use ppxlib. I think that the suggestion makes sense, I understand the benefit of locality for predictability and performance, but this is a transition that I would prefer to see done without hurry, instead of being forced to do it. You propose to add an ugly hack within ppxlib to enable a gradual transition; and in any case it's going to be your choice, or in general the choice of the people actually doing the ppxlib side of the work (not me). But to me this ugly workaround sounds inferior to finding a general solution that would work for ppx_import and also other people. (With the caveat that I barely understand the design space on the ppxlib side.)

pitag-ha commented:
> I'm not sure what "an instrumentation" is

I would say an [instrumentation](https://en.wikipedia.org/wiki/Instrumentation_(computer_programming)) is a way to analyze your program in terms of test coverage, performance etc. One example that uses that feature is `bisect_ppx` (to be applied after other PPXs).

> Making `instrument` more coherent with other ppxlib passes seems less ad-hoc, and not harder to do, than adding a "hidden register_ppx_import" function, and it may be of benefit to other users of ppxlib. Is there a reason why that would be a bad idea? Could it not be helpful to help transition other existing extensions?

We don't want `ppxlib` users to extensively use the `Before` or `After` feature for the following reason: that feature doesn't fix an exact position at which your PPX is applied. It only puts PPXs into buckets. There is a clear order of application between buckets. But PPXs inside the same bucket are simply applied in alphabetic order. So if the bucket gets big, it looses its purpose. We made clear that we don't want the `Before` or `After` feature to be used extensively by labeling them as "meant only for instrumentations". When I suggested using that feature as a hack for the `ppx_import` port, I added a clear comment that that's just a temporary solution and I thought that the feature could be used as is without extending it. But I would prefer keeping clear signs that it shouldn't be used extensively (only supports `impl` rewriters; it's named `instrument`).

Also notice: There is another feature for non-instrumentation whole-file transformations that really need to be applied before all other PPXs, called `preprocess_impl`. For example `ppx_optcomp` uses that. But there can be only one `preprocess_impl` PPX registered at once in a project! So also using that extensively is not a good idea and in the case of `ppx_import` it's not necessary.

> You propose to add an ugly hack within ppxlib to enable a gradual transition

Is it clearer now why I've proposed that ugly hack instead of proposing to extend the `Before`/`After` feature? Of course I'd only want to implement that ugly hack as a temporary solution and only if we have a nice long-term solution such as the change of syntax (of course without hurry, since we would have this temporary hack). If not, I would try to add a virtual extension node to `ppxlib` to try registering `ppx_import` as extension node rewriter even with its current syntax.

gasche commented:
One approach could be to implement the `ppx_import` logic using the`[%import type ...]` syntax, and then add an instrumentation rule (or preprocessing rule, except we can't because preprocessing rules are un-composable by choice) that simply transforms `type t = [%import ..] [@@foo]` into `[%import type t = ... [@foo]]`. That transformation is "obviously harmless", and the fact that it gets ordered with other instrumentation rules in an arbitrary way does not seem to be an issue. But we would still need the ability to instrument signatures/mlis.

pitag-ha commented:
I like that approach from the syntax point of view! We'd allow the new syntax while not breaking the projects using the old syntax. There are a couple of other things I don't like too much about that approach though.

One downside is that it doesn't improve anything from the performance point of view: even though the `ppx_import` logic would be applied together with all other context-free rules, the syntax transformation would still take up one whole AST pass on its own (even for folks using the new syntax there would be one whole identity AST pass).

Another downside is the one you've already mentioned that we'd have to extend the `instrument` feature for signatures for a PPX that's not an instrumentation but a PPX that rewrites the syntax. But it's also true what you're saying: having that simple transformation in the same bucket as instrumentations wouldn't do any real harm. I'll ask the other ppxlib folks what they think about that. (cc e.g. @NathanReb , since I think Jeremie is on vacation).

The last thing is not really a downside, but more of a theoretical remark about that approach: taking out the machinery of a PPX rewriter, in particular a whole file transformation, for a transformation that could be done once and for all with a command as simple as `sed 's/type\(.*\)\[%import:\([^]]*\)\]/type%import\1\2/g'` isn't ideal.

NathanReb commented:
I think @pitag-ha covered this pretty well but to summarize it all simply: We don't want to encourage users to write whole AST transformations, everyone should use context-free rules as much as possible. It is a much better approach to keep good performances and a clear and predictable transformation semantic.

We are aware that some rare and very specific ppx-es need to be written as whole AST passes, and we added support for those but this should remain an exception. Handling the order of transformations through priorities is a part of OMP's heritage, that we don't want as we are convinced it was a mistake for quite a while now.

`ppx_import` doesn't have to be written in such a way and that's why we don't want to add generic support for doing so.

That being said, we want to make the transition as smooth as possible both for `ppx_import` maintainers and users and we agree whatever solution we choose need to be compatible with the current syntax.
I think we established above that there is no satisfying solution using the current ppxlib API that will allow porting ppx_import without changing the syntax so we will have to add a workaround in ppxlib.
Out of of all the proposed workarounds, I think adding an exception in the context-free rules handling so that `ppx_import` can be registered as a context-free rule for expanding type declaration while using the current syntax is the best, performance and semantic-wise.
It requires us (ppxlib maintainers) to maintain a very specific exception for that purpose only so I'd suggest the following course of action:
1. We add this feature to ppxlib as a hidden or deprecated, made ppx_import only API entry
2. We make the ppx_import ppxlib port use that feature and release it
3. We add the support for the above suggested syntax in ppx_import, this should be a few more lines of ppxlib registration code so nothing hard to maintain on your side.
4. We deprecate the use of the old syntax, with a clear time schedule for its removal
5. We remove the old syntax support in ppx_import in a major release
6. We remove the corresponding feature in ppxlib in a major release

I think this solution is satisfying for everyone so let's settle on this. We don't have to rush the deprecation cycle but I think it's important to do move on with the syntax migration once the port is done and released. I'll be happy to help on the ppx_import side if that allows us to get rid of any ad-hoc weirdness in ppxlib.

gasche commented:
> The last thing is not really a downside, but more of a theoretical remark about that approach: taking out the machinery of a PPX rewriter, in particular a whole file transformation, for a transformation that could be done once and for all with a command as simple as `sed 's/type\(.*\)\[%import:\([^]]*\)\]/type%import\1\2/g'` isn't ideal.

In my experience, this is not how things work in practice. The following things will happen:
- the script is inadequate because it doesn't handle all the other cases (multiline declarations, `and t = [%import ..]`)
- there material online (READMEs, blog posts, etc.) that will not be updated and still use the wrong syntax
- maintainers of projects using ppx_import will not hear about the script anyway, because they are doing updates at their own rythm and any effort to warn them of the transition (on Discuss, whatever) will miss most of them; we could notify them individually on opam, but that is work and people use ppx_import in one-off scripts that are not on opam.
- a sizeable part of the ppx_import codebase is in research code that gets maintained on-and-off (someone decides to adopt code that was written 4 years ago and continue it), with new contributors not necessarily being familiar with the details of ppx implementation (in practice no one is). Those contributors will never see the one-or-two "transition versions" we have in mind, they will jump from OCaml 4.08 to OCaml 5.04 and everything will break at once for them.
- After the transition is done, the error message shown on "incorrect" (previously supported for years) inputs will ppxlib's generic "I don't know what to do with this extension" message, unless we specifically think of keeping an `import` extension registered on type expressions to show a proper recommendation to switch to the new syntax.

All of this can be dealt with, but it's not a one-line sed script. It never is. I would *happily* take the performance hit of a mostly-useless extra pass (only for users who decided to use ppx_import in the first place, none of which ever complained to us about any compile-time performance issue) instead of having to deal with all of this.

pitag-ha commented:
>  unless we specifically think of keeping an import extension registered on type expressions to show a proper recommendation to switch to the new syntax.

That sounds like a very good idea to me if we go down the road of the syntax change (which I still think we should)! That could also simply be implemented as a context-free rule.

> maintainers of projects using `ppx_import` will not hear about the script anyway

If we implement your suggestion of keeping an import extension for the old syntax to show an error message about how to switch to the new syntax, I think that shouldn't be a problem, should it? We could write an update script, point to the script in the README and point to the README in the error reporting for the old syntax.

> most online material (READMEs, blog posts, etc.) will not be updated and still use the wrong syntax forever

If I'm not missing anything, the only READMEs affected should be other PPX. Is that right? I've just grepped quickly through the `ppx-universe` and the only places affected should be `ppxlib`'s `HISTORY.md` and `ppx_optcomp`'s README. I'd be happy to submit the two PRs to fix that. About the blog posts: I think think it's a well-known fact that blog posts can get out-of-date. And even if people follow the out-of-date blog post suggestion of the old syntax, they would get the new syntax suggested in the error message with your trick.

> they will jump from OCaml 4.08 to OCaml 5.04 and everything will break at once for them

That's a good point. But to be fair, if you jump from OCaml 4.08 to OCaml 5.04, you probably expect quite some breakages. And again, there would be an error message making clear how to fix it.

> the script is inadequate because it doesn't handle all the other cases (multiline declarations, and `t = [%import ..]`)

Again, good point! My one-liner suggestion has been a bit too naive. Still, I'm sure we could figure out a quite simple script that would work. I also know that @NathanReb and some folks have been talking about a tool using a `dune promote` workflow to help updating to a new major version of some project (I think actually in the context of `dune`, but we could see if we could also use it here).

ejgallego commented:
Thanks folks for all the comments and work on this issue, while I tend to be in the more conservative side, and I acknowledge the issues @gasche does point out just above, I think in this case the breakage is justified, as making the ppx_import syntax compositional will save us much time in the future, at the cost of the migration.

For OPAM that's not super bad as only 10 projects depend on ppx_import, and a few of them I'm involved with .

> * the script is inadequate because it doesn't handle all the other cases (multiline declarations, `and t = [%import ..]`)

True, on the other hand for the few remaining problems, manual fix seems straightforward in this case, so not a super huge deal IMHO.

> * most online material (READMEs, blog posts, etc.) will not be updated and still use the wrong syntax forever

Indeed a big problem, tho there is not so much material covering ppx_import as in other cases, so impact is more limited.

> maintainers of projects using ppx_import will not hear about the script anyway, because they are doing updates at their own rythm and any effort to warn them of the transition (on Discuss, whatever) will miss most of them; we could notify them individually on opam, but that is work and people use ppx_import in one-off scripts that are not on opam.

I'd expect that this could be handled kinda okish, for the 1.x series, a warning would appear, so they would eventually notice it; as for the 2.x series, indeed the opam warning and release notes should do it; I'd say it is expected that the maintainers do realize that a 1.x -> 2.x transition may impact them, so they could read the CHANGES file.

> they will jump from OCaml 4.08 to OCaml 5.04 and everything will break at once for them.

I'd expect them to be able to debug the breakage in the event of such a jump, wouldn't they?

> unless we specifically think of keeping an import extension registered on type expressions to show a proper recommendation to switch to the new syntax.

I'd do indeed do this, in Coq we have done something similar [keeping parsing entries for removed stuff, so the new syntax can be recommended in an error message] and worked well.

But it is true that at some point we removed the error entries, and got some users doing jumps and not getting the recommendation. But I'd say that we can keep it for a while with ppx_import.

pitag-ha commented:
@ejgallego , awesome! 

@NathanReb is currently working on adding a virtual extension node to `ppxlib` for the temporal solution for a `ppx_import.1.x` release. Once that's done, we can finish this port and start working on the long-term solution by adding the new syntax and deprecating the old one.

tatchi commented:
Thanks for all the comments and the great feedback.

> @NathanReb is currently working on adding a virtual extension node to `ppxlib` for the temporal solution for a `ppx_import.1.x` release. Once that's done, we can finish this port and start working on the long-term solution by adding the new syntax and deprecating the old one.

Happy to try implementing that solution once @NathanReb finished his work 😁

ejgallego commented:
Great!

@gasche , any further thoughts on the tradeoff between the compat breakage vs making the syntax compositional?

I think your general points about ppx design are still valid, however it seems to me that ppx_import actually doesn't need them, so my pref is for the syntax update.

gasche commented:
I don't have a strong preference, and you are a heavier user of ppx_import than I am, so I am happy to follow your intuition. In any case, I think there is no debate that the proposed syntax is slightly better. The main point of discussion was how to manage the transition (make ppxlib more general to support the patterns we need, or hard-code a specific transition tailored for ppx_import), and this is a choice for ppxlib people rather than ppx_import people. So: yes, please go ahead!

pitag-ha commented:
> Happy to try implementing that solution once @NathanReb finishes his work :grin:

@tatchi, are you still down for that? @NathanReb has now implemented the new function `Extension.__declare_ppx_import` necessary to finish this port. He has also given quite some detail about how to use it and what to keep in mind when doing so on [the corresponding PR description](https://github.com/ocaml-ppx/ppxlib/pull/271#issue-696125992).

So if you'd like to, you could finish the port now! For now (until we cut the next release), you could just pin `ppxlib` to our main branch and work with the dev version by adding something like
```
pin-depends: [
  [
     "ppxlib.0.23.0"
     "git+https://github.com/ocaml-ppx/ppxlib#main"
  ]
]
```
to `ppx_import`'s opam file (just temporarily of course. it would have to be removed before merging this).

What do you think?

tatchi commented:
> > Happy to try implementing that solution once @NathanReb finishes his work 😁
> 
> @tatchi, are you still down for that? @NathanReb has now implemented the new function `Extension.__declare_ppx_import` necessary to finish this port. He has also given quite some detail about how to use it and what to keep in mind when doing so on [the corresponding PR description](https://github.com/ocaml-ppx/ppxlib/pull/271#issue-696125992).
> 
> So if you'd like to, you could finish the port now! For now (until we cut the next release), you could just pin `ppxlib` to our main branch and work with the dev version by adding something like
> 
> ```
> pin-depends: [
>   [
>      "ppxlib.0.23.0"
>      "git+https://github.com/ocaml-ppx/ppxlib#main"
>   ]
> ]
> ```
> 
> to `ppx_import`'s opam file (just temporarily of course. it would have to be removed before merging this).
> 
> What do you think?

Hi @pitag-ha, thanks for notifying me. I'm still down for it! Things are not very fresh in my memory though, it will probably take me some time to re-read the thread and understand what was the issue/proposed solution 😆

Anyway, I'm definitely happy to give it a shot 🤗

pitag-ha commented:
Sure! And please just ask if you have any question :)

tatchi commented:
I rebased the branch to bring in https://github.com/ocaml-ppx/ppx_import/pull/61 (the failing test) and modified the code to use the new ppx_import syntax. The relevant commit is https://github.com/ocaml-ppx/ppx_import/pull/54/commits/23fd6a8b8ff7c7f2dccad08dcc013e4c3501de6b

To make it work, I slightly had to change the `ppxlib` code in order to expose an `input_name` function in the `Expansion_context.Extension` module (see https://github.com/ocaml-ppx/ppxlib/pull/284).

For the `type_declaration`, the only thing I had to do is to write the `expand` function. Then I was able to re-use the `type_declaration` function that contains all the logic.

For the `module_type_declaration`, I had to slightly change the code. The `module_type` function was previously receiving an input of type `Ppxlib.module_type`. It receives now an input of type `Ppxlib__Import.package_type`.

The challenge I faced is that there were some code paths that were using the `module_type` received as input, modifying some of its properties and returned it. Since we no longer receive a `module_type` as input, I had now to construct one from scratch. Not sure how it matters, especially since all the tests are passing (including the one that was failing before).

tatchi commented on src/ppx_import.ml:
```diff
@@ -465,79 +465,72 @@ let rec psig_of_tsig ~subst (tsig : Compat.signature_item_407 list) :
+  if tool_name = "ocamldep" then
+    if is_self_reference ~input_name lid then
+      (* Create a dummy module type to break the circular dependency *)
+      Ast_helper.Mty.mk ~attrs:[] (Pmty_signature [])
```
That's one of the changes I was talking about. A new `module_type` is created from scratch here. We can't reuse any `attributes` from a `module_type` we receive as input anymore so I used an empty list. Could that cause any issues?

NathanReb commented:
What's the status of this PR?

I think it's based on the latest ppxlib additions and should be ready for review isn't it?

I'm currently working on some follow up changes and bug fixes on ppxlib and wanted to make sure that `ppx_import` does not modify the type declaration attributes. I'll take a look at the code to make sure and if it is indeed the case, that should be part of the contract between ppxlib and ppx_import.

If I was wrong to assume that the need to discuss how `ppx_import` modifies attributes of type declarations and how to properly interpret this in ppxlib.

tatchi commented:
> What's the status of this PR?
> 
> I think it's based on the latest ppxlib additions and should be ready for review isn't it?

From what I remember, I should have adapted the code to what is described in this PR: https://github.com/ocaml-ppx/ppxlib/pull/271#issue-696125992 (or at least from what I understood).

All the current + the failing test are passing. So yes it should be ready for a review 😁

pitag-ha commented on src/ppx_import.ml:
```diff
@@ -465,79 +465,72 @@ let rec psig_of_tsig ~subst (tsig : Compat.signature_item_407 list) :
+          let psig =
+            psig_of_tsig ~subst (List.map Compat.migrate_signature_item tsig)
+          in
+          Ast_helper.Mty.mk ~attrs:[] (Pmty_signature psig)
```
> Since we no longer receive a `module_type` as input, I had now to construct one from scratch. Not sure how it matters, especially since all the tests are passing (including the one that was failing before).

This place is what you were referring to, isn't it?

I'm not sure if that ever might happen, but if someone wrote 
```
module type Foo = [%import: ...] [@inner_attribute] [@@deriving ...]
```
then the inner attribute would get lost now. 

@ejgallego, do you think that's a problem? By the way, I'm not really sure: what's the use case of `ppx_import` in the context of module types? Is it really also for derivers (as in my example above) or is it actually something different? Or both?

ejgallego replied:
I can't tell, I'll test when I can, but I think it should be OK for my use cases.

pitag-ha replied:
Thanks! Fingers crossed.

pitag-ha replied:
So you have use cases of `ppx_import` on module types yourself?

pitag-ha replied:
@ejgallego, how is this going? Have you already had time to test this?

ejgallego replied:
Nope sorry, but will go ahead ASAP.

tatchi replied:
> This place is what you were referring to, isn't it?

Yes indeed. And also [here](https://github.com/ocaml-ppx/ppx_import/pull/54/commits/23fd6a8b8ff7c7f2dccad08dcc013e4c3501de6b#diff-2643fbce85cdaa736b4e63b3d83b2ba311bf7eb40d0ba9775a6a11f0198f0854R475) and [here](https://github.com/ocaml-ppx/ppx_import/pull/54/commits/23fd6a8b8ff7c7f2dccad08dcc013e4c3501de6b#diff-2643fbce85cdaa736b4e63b3d83b2ba311bf7eb40d0ba9775a6a11f0198f0854R478) but as you pointed out, these are just for `ocamldep` so that's probably fine ?

pitag-ha replied:
@ejgallego, kind reminder. Have you already had the chance to look into this? :)

ejgallego replied:
@pitag-ha @tatchi , sorry for the delay, I'm happy to report that pinning `ppx_import` to this PR and using OCaml 4.12.1, coq-serapi test suite seems work fine and passes the test suite! All without a single change to the source code, congrats on the amazing work!

pitag-ha reviewed (commented):
Nice, thanks for all the work @tatchi and sorry for the delay in having another look at it! By the way, you can always ping me.

> The challenge I faced is that there were some code paths that were using the `module_type` received as input, modifying some of its properties and returned it. Since we no longer receive a `module_type` as input, I had now to construct one from scratch. Not sure how it matters, especially since all the tests are passing (including the one that was failing before).

See my comment below about that.

pitag-ha commented on src/ppx_import.ml:
```diff
@@ -465,79 +465,72 @@ let rec psig_of_tsig ~subst (tsig : Compat.signature_item_407 list) :
 
-let module_type ~tool_name ~input_name (modtype_decl : Ppxlib.module_type) =
+let module_type ~tool_name ~input_name
+    (package_type : Ppxlib__Import.package_type) =
```
Hey @tatchi, have you seen this suggestion? It's just better to use the public API instead of the internal one, since there's no guarantee for the internal one to stay stable.

pitag-ha commented:
Hey @tatchi and @ejgallego, we've now released ppxlib.0.24.0 which contains all the new changes needed (see entries 2, 3 and 4 in [release notes](https://github.com/ocaml-ppx/ppxlib/releases/tag/0.24.0/)). Could you unpin ppxlib and instead bound it to `>=0.24.0`, @tatchi? And could you dig a bit into the [last open question](https://github.com/ocaml-ppx/ppx_import/pull/54#discussion_r753416624), @ejgallego? We might be close to having this mergable soon :crossed_fingers: !

ejgallego commented:
Hi folks, I could finally perform some testing on my end, and this PR seems to work very well!

I thus propose to merge and do a release if everyone is happy.

tatchi commented:
Thanks for taking the time to test it 🤗 All fine for me to merge it but,
to let you know, I noticed that it doesn't work with OCaml 4.13 (due to a
change in the AST). I've a local branch where I should have fixed that.

I was waiting for that PR to be merged before opening a new one with the
above mentioned fix. Up to you to decide if it should be included in the
next release :)

Le jeu. 6 janv. 2022 à 19:38, Emilio Jesús Gallego Arias <
***@***.***> a écrit :

> Hi folks, I could finally perform some testing on my end, and this PR
> seems to work very well!
>
> I thus propose to merge and do a release if everyone is happy.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml-ppx/ppx_import/pull/54#issuecomment-1006824662>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKV7VEEW3N7H4Y6TCLSLE3UUXORZANCNFSM446TYDEQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ejgallego commented:
> I was waiting for that PR to be merged before opening a new one with the above mentioned fix. Up to you to decide if it should be included in the next release :)

I think no problem at all; I propose to merge on Monday evening, so we can maybe cook the release end of next week.

gasche commented:
Thanks @tatchi and @pitag-ha for the hard work, and thanks @ejgallego for the review!

@ejgallego, will you take care of preparing a new release?
(This is a major change, so I think we should be careful about the build reports for our reverse dependencies, as provided by the opam-repository CI.)

ejgallego commented:
> @ejgallego, will you take care of preparing a new release?

Yup, as soon as @tatchi adds the PR for 4.13 support.

> (This is a major change, so I think we should be careful about the build reports for our reverse dependencies, as provided by the opam-repository CI.)

Sure, I think it will be 1.9.0 tho, as 2.0 may be reserved for #62 if finally goes ahead.

tatchi commented:
> Yup, as soon as @tatchi adds the PR for 4.13 support.

I will do open a PR for that later on today :)
